### PR TITLE
better auth configuration and validation; cleanup

### PR DESCRIPTION
**Pull request by jdef**

### CODE REVIEW

**Comment by jdef** [comment]
@timcharper PTAL; I ran into some problems during integration testing because I fat-fingered the `ConfigMap` init. Decided to implement some better config map handling / validation.

File: httpcli/basic/basic.go [hunk 1]
```diff
@@ -9,16 +9,26 @@ import (
 
 // Register registers a DoerFactory for HTTP Basic authentication
 func Register() {
-	httpcli.Register(httpcli.AuthBasic, httpcli.DoerFactory(func(cm httpcli.ConfigMap, c *http.Client) httpcli.Doer {
+	httpcli.Register(httpcli.AuthBasic, httpcli.DoerFactory(func(cm httpcli.ConfigMap, c *http.Client) (doer httpcli.Doer) {
```
Comment (timcharper) [inline httpcli/basic/basic.go:12]: I'm not sure I see a benefit to giving the return value a name; why not just keep it `return Doer(c, config)`?

File: httpcli/basic/basic.go [hunk 2]
```diff
@@ -9,16 +9,26 @@ import (
...
 		obj := cm.FindOrPanic(httpcli.AuthBasic)
 		config, ok := obj.(Credentials)
 		if !ok {
-			panic(fmt.Sprintf("expected Credentials instead of %#+v", obj))
+			panic(fmt.Errorf("expected Credentials instead of %#+v", obj))
 		}
-		return Doer(c, config)
+		validate(config)
+		if c != nil {
```
Comment (timcharper) [inline httpcli/basic/basic.go:19]: if c is nil then we should probably panic. Or, we should modify this function to return an error.

**Comment by timcharper** [comment]
Huh, it did work for me. Curious in which scenario it failed?

I'm glad to see there are tests. I'm struggling to see the purpose of the factory registration. It seems like excessive generality to me. The problem could simply be solved if `records.InitMesosAuthentication` returned a `Doer` func. Putting the state in a general map rather than just embedding it in the struct implementing the `Doer` interface creates more possibilities for failure. I think this is perhaps manifested by an increase in the number of panic statements; it's panicking in areas where the type system could be providing compile-time proof, which I tend to prefer.

I don't mean to be critical. I'm super happy you're spending the time to get this right, I just wonder if the complexity brought by the factory actually buys anything. Maybe it does and I just can't see it. I'm not unhappy with the factory pattern. If it works and it makes you happy, I say merge it. :) It's not a huge deal to me.

File: httpcli/basic/basic.go (hunk 2 above)
Comment (jdef) [inline httpcli/basic/basic.go:19]: the `DoerFactory` interface says that if `c == nil` then only validate. yes, it's kind of overloading the intended purpose of the factory.

File: httpcli/basic/basic.go (hunk 1 above)
Comment (jdef) [inline httpcli/basic/basic.go:12]: not a huge benefit other than maintaining a single `return` statement.

**Comment by jdef** [comment]
agreed, could be further simplified. trying to balance my time across multiple efforts and so this is where I landed here. i'm also not happy with the panics. there's certainly room for another round or two of refactoring. if we want this to land for dc/os 1.8 that refactoring may have to wait.

**Comment by jdef** [comment]
@timcharper I should also say "thanks!" for your patience w/ this process. if this works for you please merge this PR into yours.

**Comment by timcharper** [comment]
I'm happy with it is. Could polish it forever but it's concerns are nice and separated, it has tests, and it works! What more could you want? Thank you for collaborating on the change and helping to get it merged, @jdef!

I'll compile it really quick and just check thrice that it works. I'll report the success on the original pull.

**Comment by timcharper** [comment]
@jdef it has tests, it works, it's modular... I'm happy with it as is. Thank you for collaborating on it and helping to get it merged!

**Pull request merged**